src/com/palmergames/bukkit/towny/tasks/DrawSmokeTask.java

```diff
@@ -22,13 +35,45 @@ public DrawSmokeTask(Towny plugin) {
 	public void run() {
 		for (Player player : Bukkit.getOnlinePlayers()) {
 			Resident resident = TownyAPI.getInstance().getResident(player);
-			if (resident == null || !resident.hasMode("constantplotborder"))
+			if (resident == null)
+				continue;
+
+			if (resident.hasMode("constantplotborder")) {
+				WorldCoord wc = new WorldCoord(player.getWorld().getName(), Coord.parseCoord(player.getLocation()));
+				CellBorder cellBorder = BorderUtil.getPlotBorder(wc);
+				
+				cellBorder.runBorderedOnSurface(1, 2, DrawSmokeTaskFactory.showToPlayer(player, DrawSmokeTaskFactory.getAffiliationColor(resident, wc)));
 				continue;
+			}
 
-			WorldCoord wc = new WorldCoord(player.getWorld().getName(), Coord.parseCoord(player.getLocation()));
-			CellBorder cellBorder = BorderUtil.getPlotBorder(wc);
+			if (resident.hasMode("townborder")) {
+				Town town = TownyAPI.getInstance().getTown(player.getLocation());
+				if (town == null)
+					continue;
+
+				// Check if this town's cellBorders are already cached and less than 30 seconds old.
+				List<CellBorder> cellBorders = null;
+				if (townCachedTime.containsKey(town.getUUID()) &&
+					townCachedTime.get(town.getUUID()) > System.currentTimeMillis()) {
+						cellBorders = townCellBorderMap.get(town.getUUID());
+				}
+
+				// Not cached or cached List was stale.
+				if (cellBorders == null) {
+					List<WorldCoord> wcs = town.getTownBlocks().stream()
+							.map(TownBlock::getWorldCoord)
+							.filter(wc -> wc.getBukkitWorld().getName().equals(player.getWorld().getName()))
+							.filter(wc -> wc.getNormalizedDistanceFromLocation(player.getLocation()) <= 200)
+							.collect(Collectors.toList());
+					cellBorders = BorderUtil.getOuterBorder(wcs);
+					townCachedTime.put(town.getUUID(), (System.currentTimeMillis() + (long) TimeMgmt.ONE_SECOND_IN_MILLIS * 30));
+					townCellBorderMap.put(town.getUUID(), cellBorders);
+				}
+
+				cellBorders.forEach(cb -> cb.runBorderedOnSurface(1, 2, DrawSmokeTaskFactory.showToPlayer(player)));

```

Comment:
using showToPlayer that takes a player and a color would be slightly better here, since it avoids parsing the worldcoord and looking up the resident again. The color should be cached so that getAffiliationColor is only called once.

Reply:
Done

Reply:
We now only get the colour once.